src/index.js

```diff
@@ -47,7 +48,11 @@ class Dotenv {
 
     Object.keys(blueprint).map(key => {
       const value = vars.hasOwnProperty(key) ? vars[key] : env[key]
-      if (!value && safe) {
+
+      const isMissing = typeof value === 'undefined' || value === null ||
```

Comment:
Thanks again @mkrause for this, I think i'm hesitant to merge this in because readability is challenging.

Before we had `if (!value && safe)` and no we have `if (safe && isMissing)` which is a little less straitforward.

Maybe if we could set it up in such a way that stays closer to the ease in readability, like `if (!value && required)` so we can know exactly why this error would be thrown just by looking at the code.

Let me know if you have questions.

Reply:
Is the issue in the `isMissing` assignment or in the `if` statement?

Because the following seems clear to me:

`if (safe && isMissing) { ... }`

This just says "if we're in safe mode, and the variable is considered "missing", then throw the "Missing var" error". Whether a variable is considered missing depends on the `allowEmptyValues` flag, if it's true then an empty string is considered missing (in addition to the other possible falsy values, `null` and `undefined`).

Would it help if I changed the `isMissing` declaration to an `if` statement instead? Something like:

```
let isMissing = !value
if (!allowEmptyValues) {
  isMissing = isMissing && (value !== '')
}

Reply:
Hey @mkrause I haven't forgotten about you.

Real job has been busy so I haven't had time to circle back to this guy.

So i'm trying to come up with a simple psuedo-if statement that highlights what I think is most readable, and as I try to do that I realize I personally do not understand the use-case of `allowEmptyValues`. Perhaps you can help me understand it.

This is the best pseudo-if I can come up with:
```
// make sure value is set, or at least declared when allowEmptyValues is true
const isSet = value || (allowEmptyValues && value === '')
if (safe && !isSet) {
   // error
}
```

Reply:
> So i'm trying to come up with a simple psuedo-if statement that highlights what I think is most readable, and as I try to do that I realize I personally do not understand the use-case of allowEmptyValues. Perhaps you can help me understand it.

So, this is to allow environment variables to be empty without causing a failure. An example would be a configuration value for an optional feature. Maybe I have a feature `FOO` which is enabled if `FOO` is nonempty but otherwise disabled.

Another common example is where one environment variable depends on another. For example, let's say I have a flag `ANALYTICS_ENABLED` to enable/disable analytics. If enabled, we might have certain other `ANALYTICS_` variables to configure that feature. If analytics is disabled, you would like to leave those variables empty.

Currently these use cases require a workaround, like some placeholder value (I often use "~"). Besides being ugly, it's also dangerous as sometimes your placeholder may be a valid value.

> This is the best pseudo-if I can come up with

This is essentially doing the same thing, just reversed. If you prefer this I can change it.

BTW, I can try to motivate my current approach a bit more:

* I use the term "missing" instead of "set" because that's the terminology used in the error and the rest of the code.
* I write out the different cases (undefined, null, empty string) rather than using `value` or `!value` because those three cases are implicit under JS's truthiness rules, so I figured since we make a distinction between empty string and the other falsy values it would be more readable to write them out explicitly.

But I'm not adamant about these choices of course. :) Would you like me to change it to your pseudocode instead?

Reply:
No lets discuss more, I still think my understanding is lacking of the purpose of the property.

In your example of the `ANALYTICS_ENABLED` sounds like you could just treat it as a flag right?

```
# .env
ANALYTICS_ENABLED=sure
```

and in js:
```js
if (process.env.ANALYTICS_ENABLED) {
  // do something
}
```

...This would also work as expected in your use-case for checking other variables

```js
const enabled = process.env.ANALYTICS_ENABLED

if (enabled) {
  if (process.env.ANALYTICS_SOMETHING_ELSE) {
    // do other things
  }
}
```

... The code just checks for non falsy requests, which is fine. JS has flexibilities to use falsy to our advantage here and if you want it to be false, just don't set it right?

I guess this is where I get hung up. If you want your env variable to be true, just set it to anything, if you want it to false, don't set it. By adding the "explicitly set but to no value" approach I find myself confused at the additional dynamic and complexity we are adding to simple env variables.

Either I'm super-missing the point, or I believe env variables are overloaded in the purpose.

Please help fill in the gaps if you can @mkrause.

Reply:
When safe mode is enabled, and the variable is empty, webpack won't compile at all. The build will fail immediately because dotenv-webpack notices a missing variable. So checking the presence of the variable in the code won't work.

Reply:
Am I crazy to think it should just be removed from the `.env.example` file?

Reply:
That's also a workaround, yeah. It doesn't seem like the right solution to me though. You're essentially performing two workarounds:

1. Leaving the variable out of the "schema" even though it does exist (it just might be empty).
2. Performing additional logic in the code to consider the nonexistence of a variable as though it were there (just empty).

Another reason I wouldn't like to do it that way is because it means that when the variable gets added to an existing project, all existing users of the project will not be notified that there's a new variable. The code will just assume the value is empty. Whereas with the `allowEmptyValues` approach the user has to be make an explicit choice what to fill in the next time they start the application. That's my personal preference though.

Reply:
@mrsteele Final verdict on this PR?

Reply:
@mkrause You have a background in typescript/typed languages (Java)?

I'm so torn on this. On the one hand I like explicit architectures, but the "allowEmptyVariables" thing has always tripped me up.

Aside from dotenv-safe, can you find other plugins/systems that use this "explicitely-defined-empty approach"?

Reply:
What are your thoughts on the falsy principles in js? I'm personally a big fan, and I think it helps with readability and architecture design philosophies.

Changing a few names around:
```js
// getFoo has return type `null | string`
const options = { path: getPath() || __dirname };
if (!options.path) {
   throw new Error('You have no path!')
}
```

In my above example, we have used falsy principles to be intentionally inclusive of potential problem-values. The error would be thrown if `path` is `null`, `undefined`, `0`, `false` or even `""`. Checking explicitly for `null` is restrictive and potentially problematic when you forget about the other falsy values. With your developmental architecture you could force the data to flow in such a way to only allow `string | null` but you now have to cary the weight of remember you used `null` instead of any of the other falsy values.

Reply:
I think it's a double-edged sword. I make use of the kinds of shortcuts you mention if it's clear what's happening. For example, if we have a variable of type `null | object` then using `foo || defaultValue` is a good idiomatic way to handle defaults.

If on the other hand it's not immediately obvious (e.g. a variable has a complex type like `undefined | null | string | string[]` or something), then I'll handle the different cases explicitly. Relying too much on JS's weak typing rules is an easy way to get bitten by bugs, I know from experience.